test/functional/conftest.py

```diff
@@ -1,78 +1,68 @@
 # Note that this file is special in that py.test will automatically import this file and gather
 # its list of fixtures even if it is not directly imported into the corresponding test case.
-import pathlib
 
+import pathlib
 import pytest
+import sys
+import typing as ty
 
 import ipfshttpclient
 
 
 TEST_DIR = pathlib.Path(__file__).parent
 
 
-__is_available = None
-def is_available():  # noqa
+@pytest.fixture(scope='session')
+def ipfs_is_available() -> bool:
 	"""
 	Return whether the IPFS daemon is reachable or not
 	"""
-	global __is_available
 
-	if not isinstance(__is_available, bool):
-		try:
-			ipfshttpclient.connect()
-		except ipfshttpclient.exceptions.Error:
-			__is_available = False
-		else:
-			__is_available = True
+	try:
+		with ipfshttpclient.connect():
+			pass
+	except ipfshttpclient.exceptions.Error as e:
+		print('\nFailed to connect to IPFS client', file=sys.stderr)
+		print(e, file=sys.stderr)
 
-	return __is_available
+		return False
+	else:
+		return True
 
 
 def sort_by_key(items, key="Name"):
 	return sorted(items, key=lambda x: x[key])
 
 
-def get_client(offline=False):
-	if is_available():
-		return ipfshttpclient.Client(offline=offline)
+def _generate_client(
+		ipfs_is_available: bool,
+		offline: bool
+) -> ty.Generator[ipfshttpclient.Client, None, None]:
+	if ipfs_is_available:
+		with ipfshttpclient.Client(offline=offline) as client:
+			yield client
 	else:
 		pytest.skip("Running IPFS node required")
 
 
 @pytest.fixture(scope="function")
-def client():
-	"""Create a client with function lifetimme to connect to the IPFS daemon.
-
-	Each test function should instantiate a fresh client, so use this
-	fixture in test functions."""
-	with get_client() as client:
-		yield client
+def client(ipfs_is_available):
+	yield from _generate_client(ipfs_is_available, False)
 
 
 @pytest.fixture(scope="function")
 def offline_client():
-	"""Create a client in offline mode with function lifetimme"""
-	with get_client(offline=True) as client:
-		yield client
+	yield from _generate_client(ipfs_is_available, True)
 
 
 @pytest.fixture(scope="module")
 def module_client():
-	"""Create a client with a module lifetime to connect to the IPFS daemon.
-
-	For module-scope fixtures that need a client, if the client is to be created
-	automatically using a fixture (to keep client creation code centralized
-	here), that client-creating fixture must also be module-scope, so use
-	this fixture in module-scoped fixtures."""
-	with get_client() as client:
-		yield client
+	yield from _generate_client(ipfs_is_available, False)
 
 
 @pytest.fixture(scope="module")
 def module_offline_client():
-	"""Create a client in offline mode with module lifetime."""
-	with get_client(offline=True) as client:
-		yield client
+	yield from _generate_client(ipfs_is_available, True)
```

Comment:
How do these work given that they don't actually depend on the `ipfs_is_available` fixture?

Reply:
Long form: https://docs.pytest.org/en/stable/fixture.html

tldr:

The style used for the IPFS client is a generator style. For example, for a fixture with `function` scope (`scope='function' - the default), when a test runs, pytest will start the fixture generator first, then at the point the generator `yield`s, pytest will call the test function, passing the fixture as an argument. When the test function completes, control is returned back to the fixture at the yield statement where it continues where it left off.

This style is useful for fixtures that provide values requiring cleanup (database connections, file handles, IPFS server connections, etc.).

From a test function, the fixture value is accessed by declaring an argument with the same name as the fixture function. For example:

    @pytest.fixture(scope='module')
    def foo():
        return 'some value used by the tests'

    def test_bar(foo):
        assert foo = 'some value used by the tests'

Reply:
Thanks, I knew that, but how does a fixture like this:

```
@pytest.fixture(scope="module")
def module_offline_client():
	"""Create a client in offline mode with module lifetime."""
	yield from _generate_client(ipfs_is_available, True)
```

… work if it *uses* the name `ipfs_is_available`, but doesn't actually take `ipfs_is_available` an argument/fixture-name.

Thinking about it again, I think it uses fixture definition `ipfs_is_available` from global scope, which, as a function, is always truethy. So it will unconditionally connect to the daemon, even when it is not actually available.

I think this is a bug and not some pytest magic I just didn't understand like I first assumed. :grin:

Reply:
In the `foo/bar` example above, the `foo` argument of `test_bar` is a `bool` not a `Callable` - `pytest` evaluates the fixture (in this case, by simply calling it, because `foo` is not a generator, but a plain `Callable`), and then pytest will pass the return value to whatever tests are in scope (in this case, any test in the same module as the first test in a module that used the fixture, due to `scope='module'`).

Reply:
Yes, but…

Reply:
I love the dramatic pause. It made me feel really good about myself when I saw the problem.

Thanks for that :smile: